[ews] Rebuilding for retries should be done on builders instead of testers especially for macOS Sequoia

aj062 opened the pull request:
#### 4102038712b11fc379e622eb704e01157018d485
<pre>
[ews] Rebuilding for retries should be done on builders instead of testers especially for macOS Sequoia
<a href="https://bugs.webkit.org/show_bug.cgi?id=201935">https://bugs.webkit.org/show_bug.cgi?id=201935</a>
<a href="https://rdar.apple.com/49789674">rdar://49789674</a>

Reviewed by Brianna Fan and Elliott Williams.

* Tools/CISupport/ews-build/steps.py:
(CompileWebKit.evaluateCommand):
(RunWebKitTests.evaluateCommand):
(ReRunWebKitTests.evaluateCommand):
(ReRunAPITests.doOnFailure):
* Tools/CISupport/ews-build/steps_unittest.py: Added unit-tests.
* Tools/CISupport/Shared/steps.py: Added build step WaitForDuration.
* Tools/CISupport/Shared/steps_unittest.py: Added unit-tests.

Canonical link: <a href="https://commits.webkit.org/306099@main">https://commits.webkit.org/306099@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2e66c22cd3f47aa4d69437dd178b777967231f26

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/140352 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/12733 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/1865 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/148672 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/93439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/142225 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/13446 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/12887 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/107589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/93439 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/143303 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/10362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/125649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/88485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/9983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/7526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/8785 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/119222 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/167/builds/1659 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/151300 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/12421 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/169/builds/1729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/115890 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/139775 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/12435 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/10642 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/116225 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/11365 "Passed tests") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/122133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/67434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21656 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/12464 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/1589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/12204 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/12402 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/12248 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -82,6 +82,9 @@
 SHOULD_FILTER_LOGS = load_password('SHOULD_FILTER_LOGS', default=True)
+SUFFIX_WITHOUT_CHANGE = '-without-change'
+MACOS_SONOMA_TRIGGER = 'macos-sonoma-release-build-ews'
+MACOS_SONOMA_BUILDER_NAME = 'macOS-Sonoma-Release-Build-EWS'
```
This is going out of date now that we are dropping support for Sonoma.

What's preventing us from making this OS independent, and killing `CompileWebKitWithoutChange` on all testers? Just hardware resources, or something else?

aj062 replied:
Yes, hardware resources would be the primary deciding factor for killing `CompileWebKitWithoutChange` on all testers. We can enable it on one queue for now, see how it goes (especially performance), and based on that decide if we want to expand it to every queue.

Updated the PR for Sequoia.

emw-apple replied:
Sequoia has debug and release builders, so I think we need to match both, right? I'm not sure I see the value in migrating one configuration before the other, and we need to do both.

emw-apple reviewed (commented):
I know this is still a draft but wanted to give quick feedback to make sure we are on the same page.

One thing that we will still need to do is start setting the deployment target in the `build-webkit` invocation provided by CompileWebKit. Eventually all builders will run the same OS version, and the deployment target is what will determine which OS they build for. For example, the Sequoia build queue needs to run:

```
Tools/Scripts/build-webkit ... MACOSX_DEPLOYMENT_TARGET=15.4
```

So there will need to be a table of some kind that maps from queue name to deployment target version.

emw-apple commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -3417,6 +3422,16 @@ def evaluateCommand(self, cmd):
+                            RevertAppliedChanges(),
+                            ValidateChange(verifyBugClosed=False, addURLs=False),
+                            CompileWebKitWithoutChange(),
+                            GenerateS3URL(f"{self.getProperty('fullPlatform')}-{self.getProperty('archForUpload')}-{self.getProperty('configuration')}{SUFFIX_WITHOUT_CHANGE}"),
+                            UploadFileToS3(f"WebKitBuild/{self.getProperty('configuration')}.zip", links={self.name: 'Archive without change'}),
```
Is there any opportunity here to share the same archive between multiple builds? I'm guessing it wouldn't be easy, but seems like an obvious next step to save compute / storage.

aj062 replied:
I had some discussion about such optimization with Ryan and Alexey. Few things to note:
- EWS rebases on ToT and there are so many commits in the repo during the business hours, that most of the time EWS builds are building on different revisions, so such optimization won't help. During the off-hours (e.g.: at night), such optimization would help, but bots might be idle during that time anyways.
- There will be a lot of corner cases to handle (e.g.: considering that archiving+uploading can take minutes, and build master can be restarted anytime)

The benefit might be minimal. So if we really want that optimization, we should look into that separately. For now, maybe we can go without such optimization and see if we feel the need.

briannafan commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -5629,6 +5662,86 @@ def getResultSummary(self):
+                self.build.buildFinished([self._summary], FAILURE)
+                return defer.returnValue(FAILURE)
+
+        self._summary = f'Parent build failed with result: {Results[parent_build_result]}'
```
I thought I had was it'd be nice to link to the parent build here, but I suppose we do have the "triggered by" button at the top of the page. Perhaps it would be nice for people who aren't familiar with it (or forgot like me), but up to your discretion.

aj062 replied:
will skip it for now. It should be rare to hit this scenario of parent build failing, so the value addition is minimal.

briannafan reviewed (approved):
LGTM! It would be nice to have unit tests in this PR rather than a separate one if you can.

emw-apple commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -82,6 +82,9 @@
 SHOULD_LOAD_CONTRIBUTORS_FROM_NETWORK = load_password('SHOULD_FILTER_LOGS', default=True)
+SUFFIX_WITHOUT_CHANGE = '-without-change'
+MACOS_SEQUOIA_TRIGGER = 'macos-sequoia-release-build-ews'
+MACOS_SEQUOIA_BUILDER_NAME = 'macOS-Sequoia-Release-Build-EWS'
```
We will certainly need to support multiple triggers before too long (for instance, when Sequoia and Tahoe are both downlevel builds), so it may make sense to land this as a list of trigger/builder names. Feel free to punt, though.

aj062 replied:
Make sense. Will do that in separate PR.

aj062 committed:
[ews] Rebuilding for retries should be done on builders instead of testers especially for macOS Sequoia
https://bugs.webkit.org/show_bug.cgi?id=201935
rdar://49789674

Reviewed by Brianna Fan and Elliott Williams.

* Tools/CISupport/ews-build/steps.py:
(CompileWebKit.evaluateCommand):
(RunWebKitTests.evaluateCommand):
(ReRunWebKitTests.evaluateCommand):
(ReRunAPITests.doOnFailure):
* Tools/CISupport/ews-build/steps_unittest.py: Added unit-tests.
* Tools/CISupport/Shared/steps.py: Added build step WaitForDuration.
* Tools/CISupport/Shared/steps_unittest.py: Added unit-tests.

Canonical link: https://commits.webkit.org/306099@main
